io/src/main/scala/fs2/io/internal/InputOutputBuffer.scala

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2.io.internal
+
+import java.io.{InputStream, OutputStream}
+import java.util.concurrent.Semaphore
+
+/** Thread safe circular byte buffer which pipes a [[java.io.OutputStream]]
+  * through a [[java.io.InputStream]] in a memory efficient manner, without
+  * copying bytes unnecessarily.
+  *
+  * @note As per the interfaces of the [[java.io]] classes, all of the
+  * operations are blocking in nature and extra care should be taken when using
+  * the exposed input/output streams. Thread safety is ensured by
+  * synchronizing on individual objects of this class.
+  *
+  * This is, in spirit, a clean room reimplementation of the
+  * [[java.io.PipedInputStream]] and [[java.io.PipedOutputStream]] pair of
+  * classes which can be used to achieve similar functionality, without the
+  * thread bookkeeping which is confusing in a multi threaded environment like
+  * the effect systems in which this code runs.
+  *
+  * @param capacity the capacity of the allocated circular buffer
+  */
+private[io] final class InputOutputBuffer(private[this] val capacity: Int) { self =>
+
+  private[this] val buffer: Array[Byte] = new Array(capacity)
+
+  private[this] var head: Int = 0
+  private[this] var tail: Int = 0
+
+  private[this] var closed: Boolean = false
+
+  private[this] val readerPermit: Semaphore = new Semaphore(1)
+  private[this] val writerPermit: Semaphore = new Semaphore(1)
```

Comment:
Could you add a summary of how these permits are used? Are their values every > 1?

Reply:
I believe they can be, they are not completely foolproof, they are just here to prevent threads from looping endlessly when there is nothing to be read from the buffer. This is fine, because there is one more monitor, the `self.synchronized` block, which actually guards the critical section. Does this make any sense?

Reply:
Okay cool. My only concern is that some strange access pattern could result in `writerPermit` getting very large, which then causes lots of looping when waiting to write.

Reply:
Ok, let's not rush this, I will reiterate on the code.

Reply:
I implemented a custom synchronizer that allows at most 1 permit at any time.

Reply:
Looks great!